app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileAdapter.kt

```diff
@@ -95,49 +98,39 @@ class PinnedTileAdapter(
         iconView.layoutParams = layoutMarginParams
     }
 
-    /**
-     * takes in the home tiles cache and updates the adapter's data source
-     * and UI accordingly, assuming only one new tile is added
-     */
-    fun updateAdapterSingleInsertion(pinnedTiles: MutableList<PinnedTile>) {
-        if (pinnedTiles.size == tiles.size) {
-            // The lists must not be the same size in order
-            // for an insertion to be valid
+    fun setTiles(newTiles: MutableList<PinnedTile>) {
+        if (itemCount == 0) {
+            tiles = newTiles
+            notifyDataSetChanged()
             return
         }
-        for ((index, tile) in tiles.withIndex()) {
-            // Due to insertion, the inserted tile will be
-            // the first tile that will not match the
-            // previous list of tiles
-            if (tile != pinnedTiles[index]) {
-                tiles = pinnedTiles
-                notifyItemInserted(index)
-                return
+
+        // DiffUtil allows diff calculation between two lists and output a list of update
+        // operations that converts the first list into the second one
+        val diff = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
```

Comment:
I think we can simplify this by using `RecyclerView.widget.ListAdapter`, which uses `DiffUtil` nder the hood.